src/division.jl

```diff
@@ -106,11 +116,113 @@ function Base.divrem(f::APL{T}, g::AbstractVector{<:APL{S}}; kwargs...) where {T
     q, r
 end
 
+function isolate_variable(poly::APL, var::AbstractVariable)
+    T = typeof(substitute(Subs(), zeroterm(poly), (var,) => (1,)))
+    dict = Dict{Int,Vector{T}}()
+    for t in terms(poly)
+        d = degree(t, var)
+        dict_ts = get(dict, d, nothing)
+        ts = dict_ts === nothing ? T[] : dict_ts
+        push!(ts, subs(t, (var,) => (1,)))
+        if dict_ts === nothing
+            dict[d] = ts
+        end
+    end
+    return polynomial([
+        term(polynomial(ts), var^d) for (d, ts) in dict
+    ])
+end
+
+function coefficients_gcd(poly::APL{T}) where {T}
+    return reduce(gcd, coefficients(poly), init=zero(T))
+end
+
+"""
+    function Base.gcd(p1::AbstractPolynomialLike{T}, p2::AbstractPolynomialLike{S}) where {T, S}
+
+Returns the greatest common divisor of `p1` and `p2`.
+
+# Implementation notes
+
+The classical algorithm for computing the `gcd`, commonly referred to as the
+Euclidean Algorithm is to use a recursion with the base case `gcd(p, 0) = p`
+and the relation `gcd(p1, p2) = gcd(p2, rem(p1, p2))`.
+The relation comes from the Euclidean division:
+`p1 = q * p2 + r`,
+if `g` divides `p1` and `p2` then it divides `r` and
+if `g` divides `r` and `p2` then it divides `p1`.
+
+For multivariate polynomials, you may have `rem(p1, p2) = p1` hence this will not
+terminate.
+To ensure we make progress, we can pick a given variable `xi` and try to find
+`q1` and `q2` such that
+`q2 * p1 = q1 * p2 + r`
+and the degree of `r` in `xi` is strictly smaller than the degree of `p1` in `xi`.
+Note that if `g` divides `p1` and `p2` then it divides `r` but
+if `g` divides `r` and `p2` then it might divide `q2` and not `p1`.
+So what do we do ?
+Let `dj` be the degree of `pj` in `xi`.
+Suppose we pick `qj` to be the coefficient of `pj` in `xi^dj`.
+If `g` divides `q2` then it means that the degree of `g` in `xi` is zero.
+Therefore, if it divides `p2` then it also divides the coefficients
+of `p2` in `xi^k` for `k = 0, 1, ..., d2`.
+This means that if we ensure that these are relatively prime then we won't have
+any issue.
+So we start by computing the `gcd` `gj` of the coefficients in each degree of
+`xi` of `pj`.
+And then we compute `_gcd(p1 / g1, p2 / g2) * gcd(g1, g2)` where we can use the
+recursion `_gcd(p1, p2) = _gcd(p2, q2 * p1 - q1 * p2)` where `q1, q2` are as
+defined above. The base case is that `_gcd(p1, p2)` redirects to `gcd(p1, p2)`
+when the degree of both `p1` and `p2` is zero for `xi`. Then we pick another
+variable and we continue until we arrive at `gcd(p, 0)`.
+"""
 function Base.gcd(p::AbstractPolynomialLike{T}, q::AbstractPolynomialLike{S}) where {T, S}
+    p_vars = effective_variables(p)
+    q_vars = effective_variables(q)
+    if length(p_vars) == length(q_vars) == 1 && first(p_vars) == first(q_vars)
+        return univariate_gcd(p, q)
+    else
+        vars = p_vars ∩ q_vars
```

Comment:
Is there a consistent variable order guaranteed here? The result can be different for different variable orderings, so vars probably should always have the variables in the correct order.

Reply:
So you mean we should use `maximum` instead of `first`?